OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs

```diff
@@ -0,0 +1,47 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Actors possessing this trait should define the GainsExperience trait beforehand. When the prerequisites are fulfilled, ",
```

Comment:
This is not consistent with the code below, which cautiously checks if the actor has `GainsExperience`.
Also "actor should define trait X" is handled by `Requires<>` (which you are missing), and "beforehand" is handled by our actor initializer, which initializes the traits in order (required traits get initialized first).
Lastly you are doing the trait lookup in `AddedToWorld()`, not in the constructor, so at that point all traits are initialized (meaning you don't care about the order).

To sum up: Add `Requires<>`, change `.TraitOrDefault()` to `.Trait()` and skip the null check.